keps/sig-autoscaling/20191111-autoscaling-with-multiple-thresholds.md

```diff
@@ -0,0 +1,202 @@
+
+### Non-Goals
+
+- This does not aim at replacing the HPA Controller. 
+- Not addressing the features spec'ed in the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md).
+
+## Proposal
+
+Getting the tolerance parameter (`--horizontal-pod-autoscaler-tolerance`) from the controller level to the resource level is a good start, and has been suggested in the past (see [#39090](https://github.com/kubernetes/kubernetes/issues/39090#issuecomment-466398426)). 
+We would like to suggest going beyond this and be able to specify two thresholds. A high bound and a low bound, which extend the idea of a single threshold +/- the tolerance.
+
+There is some overlap between this KEP and the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md), we (Datadog) are heavily relying on the features they introduced and strongly believe they should be made generally available.
+
+### User Stories
+
+We have instrumented the custom controller we use internally to expose metrics, so we can best illustrate the use cases.
+
+#### Canonical use case of the bounds to control the primary scaling condition.
+
+In this example we are using the following Spec configuration:
+```
+    minReplicas: 4
+    maxReplicas: 9
+    metrics:
+    - external:
+        highWatermark: 400m
+        lowWatermark: 150m
+        metricName: custom.request_duration.max
+        metricSelector:
+          matchLabels:
+            kubernetes_cluster: mycluster
+            service: billing
+            short_image: billing-app
+      type: External
+    tolerance: 0.01
+```
+
+Starting with the thresholds, the value of the metric collected (`watermarkpodautoscaler.wpa_controller_value`) in purple when between the bounds (`watermarkpodautoscaler.wpa_controller_low_watermark` and `watermarkpodautoscaler.wpa_controller_high_watermark`) will instruct the controller to not trigger a scaling event. 
+
+We can use the metric `watermarkpodautoscaler.wpa_controller_restricted_scaling{reason:within_bounds}` to verify that it is indeed restricted. (Nota: the metric was multiplied by 1000 in order to make it look more explicit on the graph that within the red areas, no scaling event could have been triggered by the controller).
+<img width="1528" alt="Within Watermarks" src="https://user-images.githubusercontent.com/7433560/63385633-e1a67400-c390-11e9-8fee-c547f1876540.png">
+
+#### Restriction of the scaling events.
+
+This section overlaps with the [Configurable Scale Velocity KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-autoscaling/20190307-configurable-scale-velocity-for-hpa.md).
```

Comment:
I think you should just drop the parts that overlap with the previous KEP because it's been approved, and the [API](https://github.com/kubernetes/kubernetes/pull/74525) is approved and will be landing soon.